p2p/src/sync/peer_v1.rs

```diff
@@ -616,48 +612,44 @@ where
         }
 
         // The first header must be connected to a known block (it can be in
-        // the chainstate, pending_headers or requested_blocks).
+        // the chainstate or requested_blocks).
         let first_header_prev_id = *headers
             .first()
             // This is OK because of the `headers.is_empty()` check above.
             .expect("Headers shouldn't be empty")
             .prev_block_id();
 
+        // Note: we require a peer to send headers starting from a block that we already have
+        // in our chainstate or from one that we've already requested from the peer.
+        // I.e. peers shouldn't track what block headers they've sent us already and use
+        // the last header (best_sent_block_header) as a starting point for future HeaderList
+        // updates.
+        // This restriction is needed to prevent malicious peers from flooding the node with
+        // headers, potentially exhausting the node's memory.
+        // The downside of this is that the peer may have to send the same headers multiple times.
+        // So, to avoid extra traffic, an honest peer should't send header updates when the node
+        // is already downloading blocks. But still, the node shouldn't punish the peer for
+        // doing so, because it's possible for it to do so on accident, e.g. a "new tip" event
+        // may happen on the peer's side after it has sent us the last requested block but
+        // before we've asked it for more.
+
         let first_header_is_connected_to_chainstate = self
             .chainstate_handle
             .call(move |c| Ok(c.get_gen_block_index(&first_header_prev_id)?))
             .await?
             .is_some();
 
-        let first_header_is_connected_to_pending_headers = {
-            // If the peer reorged, the new header list may not start where the previous one ended.
-            // If so, the non-connecting old headers are now considered stale by the peer, so
-            // we should remove them from pending_headers.
-            while let Some(known_header) = self.incoming.pending_headers.last() {
-                if known_header.get_id() == first_header_prev_id {
-                    break;
-                }
-
-                self.incoming.pending_headers.pop();
-            }
-
-            !self.incoming.pending_headers.is_empty()
-        };
-
         let first_header_is_connected_to_requested_blocks = first_header_prev_id
             .classify(&self.chain_config)
             .chain_block_id()
             .and_then(|id| self.incoming.requested_blocks.get(&id))
             .is_some();
 
         if !(first_header_is_connected_to_chainstate
-            || first_header_is_connected_to_pending_headers
             || first_header_is_connected_to_requested_blocks)
```

Comment:
Please remove this. The only source of truth is chainstate.

Reply:
Same applies to v2, btw.

Reply:
I've made this change in V1 too. We've never actually used this "optimization opportunity" when sending headers ourselves, so it shouldn't break anything.